opengovsg/FormSG
# fix: Email domain validation should be case-insensitive (#2578)

## Context

Allow submission of a form from the public if the email domain is in all caps (JOHN_DOE@FOO.BAR) and mixed case.(xXx_joHn_DoE_xXx@fOo.bAr)

Because allowed email domains have not been stored lowercased, we have to assume some fields definition in store may have "incorrect" case (mixed case, uppercase, etc.). 

We want a solution that introduces case-insensitive checks on the allowed domains, that is backward compatible, without having to backfill form definitions to correct cases.

Closes [#2578]

Takes over PR[#2622]

## Approach

normalize comparison operands to lowercase when validating allowed email domains, both in frontend and backend validators

Note:
* [`lowercase: true`](https://mongoosejs.com/docs/5.x/docs/schematypes.html#string-validators) is additionally added to [`models/field/emailField.ts`](https://github.com/opengovsg/FormSG/blob/102fbc3acf1a63e297328ca2760175b929daaa48/src/app/models/field/emailField.ts#L47-L53) to ensure that allowed domains will be stored lowercased for future forms. No attempt at correcting older forms is made.



**Breaking Changes** 

**Bug Fixes**:

- Validate emails against allowed domains in a case-insensitive manner

## Before & After Screenshots

with field settings:
![image](https://user-images.githubusercontent.com/935223/130889384-6a1aa72c-9c36-40c9-855c-19ae34d146ea.png)

**BEFORE**:
Email domain is rejected:
![image](https://user-images.githubusercontent.com/935223/130889546-b22fd24c-0856-46f1-9c91-a2e52adaca90.png)

**AFTER**:
Email domain is allowed:
![image](https://user-images.githubusercontent.com/935223/130889435-85fcb95a-4b50-4dc7-80fb-d80f43bb4578.png)


## Tests
1. Create a form with a verifiable email field and supply 2 allowed domains `foo.com` and `BAR.COM`
2. Open form, and verify that
    a. invalid domain is rejected: `me@ogp.sg` should show error message: `Not an allowed email domain`
    b. following emails are all allowed:
         - me@foo.com
         - me@FOO.com
         - me@foO.CoM
         - me@BAR.COM
         - me@bar.com
         - me@BaR.CoM
